specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+# Multithreaded Cannon Fault Proof Virtual Machine
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+  - [New Features](#new-features)
+    - [Multithreading](#multithreading)
+    - [Robustness](#robustness)
+- [Multithreading](#multithreading-1)
+  - [Thread Management](#thread-management)
+  - [Thread Traversal Mechanics](#thread-traversal-mechanics)
+    - [Thread Preemption](#thread-preemption)
+  - [Wakeup Traversal](#wakeup-traversal)
+  - [Exited Threads](#exited-threads)
+  - [Waiting Threads](#waiting-threads)
+  - [Voluntary Preemption](#voluntary-preemption)
+  - [Forced Preemption](#forced-preemption)
+- [FPVM State](#fpvm-state)
+  - [State](#state)
+  - [State Hash](#state-hash)
+  - [Thread State](#thread-state)
+  - [Thread Hash](#thread-hash)
+  - [Thread Stack Hashing](#thread-stack-hashing)
+- [Memory](#memory)
+  - [Heap](#heap)
+- [Delay Slots](#delay-slots)
+- [Syscalls](#syscalls)
+  - [Supported Syscalls](#supported-syscalls)
+  - [Noop Syscalls](#noop-syscalls)
+- [I/O](#io)
+  - [Standard Streams](#standard-streams)
+  - [Hint Communication](#hint-communication)
+  - [Pre-image Communication](#pre-image-communication)
+    - [Pre-image I/O Alignment](#pre-image-io-alignment)
+- [Exceptions](#exceptions)
+- [Security Model](#security-model)
+  - [Compiler Correctness](#compiler-correctness)
+  - [Compiler Assumptions](#compiler-assumptions)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+This is a description of the second iteration of the Cannon Fault Proof Virtual Machine (FPVM).
+When necessary to distinguish this version from the initial implementation,
+it can be referred to as Multithreaded Cannon (MTCannon).  Like the original Cannon implementation,
+the MTCannon FPVM emulates a minimal Linux-based system running on big-endian 32-bit MIPS32 architecture.
```

Comment:
Emphasize that it's a _uniprocessor_  Linux-based system. The most frequent question I get regarding MT-Cannon is whether it executes threads in parallel.